crates/validation/src/collection.rs

```diff
@@ -400,7 +426,123 @@ pub fn walk_selector(
     }
 }
 
+fn extend_read_bundle(
+    read_bundle: &models::Schema,
+    write_bundle: &models::Schema,
+    inferred_bundle: Option<&models::Schema>,
+) -> models::RawValue {
+    use json::schema::keywords;
+    use serde_json::{value::to_raw_value, Value};
+    type Skim = BTreeMap<String, models::RawValue>;
+
+    let mut read_schema: Skim = serde_json::from_str(read_bundle.get()).unwrap();
+    let mut read_defs: Skim = read_schema
+        .get(keywords::DEF)
+        .map(|d| serde_json::from_str(d.get()).unwrap())
+        .unwrap_or_default();
+
+    // Add a definition for the write schema if it's referenced.
+    // We cannot add it in all cases because the existing `read_bundle` and
+    // `write_bundle` may have a common sub-schema defined, and naively adding
+    // it would result in an indexing error due to the duplicate definition.
+    // So, we treat $ref: flow://write-schema as a user assertion that there is
+    // no such conflicting definition (and we may produce an indexing error
+    // later if they're wrong).
+    if read_bundle.get().contains(super::REF_WRITE_SCHEMA_PATTERN) {
+        let mut write_schema: Skim = serde_json::from_str(write_bundle.get()).unwrap();
+
+        // Set $id to "flow://write-schema".
+        _ = write_schema.insert(
+            keywords::ID.to_string(),
+            models::RawValue::from_value(&Value::String(REF_WRITE_SCHEMA_URL.to_string())),
+        );
+        // Add as a definition within the read schema.
+        read_defs.insert(
+            REF_WRITE_SCHEMA_URL.to_string(),
+            to_raw_value(&write_schema).unwrap().into(),
+        );
+    }
+
+    // Add a definition for an inferred schema if it's provided.
+    // Note that we previously filtered the set of retrieved schemas to those
+    // having a read schema matching super::REF_INFERRED_SCHEMA_PATTERN.
+    if let Some(inferred_bundle) = inferred_bundle {
```

Comment:
What happens if there is no inferred schema yet because this is a new collection? For example, what would happen if I published:

```yaml
collections:
  acmeCo/some/new/collection:
    writeSchema:
      properties:
        id: { type: string}
      required: [id]
    readSchema:
      allOf:
        - $ref: flow://write-schema
        - $ref: flow://inferred-schema
    key: [/id]
```

It seems like it'd fail to resolve the inferred schema ref. I'm wondering if we should use some default placeholder (`true`?) if there isn't an inferred schema yet?

Reply:
The intent is definitely to use a placeholder `{}` (equivalent to `true`). I definitely had it wired up that way, and confirmed it works as intended in end-to-end testing, but now I'm doubting myself and wondering if i screwed up a refactor. I'll double-check